From 0c8a2a5484066cb5e16ea1f2269f72a9d02da05a Mon Sep 17 00:00:00 2001
From: Locus Wei-Han Chen <locus84@andestech.com>
Date: Mon, 17 Mar 2025 15:55:08 +0800
Subject: [PATCH 166/168] andes: amp: fixed the extra section generated by
 toolchain (#263)

The new version of the toolchain adds an extra .bss section, which the
kernel can't handle causing errors.
The factor has been considered and resolved.

Reference:
Bugzilla 33773(http://e-andes.andestech.com/bugzilla5/show_bug.cgi?id=33773)

Signed-off-by: Locus Wei-Han Chen <locus84@andestech.com>

Reviewed-on: https://gitea.andestech.com/RD-SW/linux/pulls/263
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Reviewed-by: Charles Ci-Jyun Wu <dminus@andestech.com>
Reviewed-by: Mina Hui-Min Chou <minachou@andestech.com>
---
 drivers/remoteproc/andes_remoteproc.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/remoteproc/andes_remoteproc.c b/drivers/remoteproc/andes_remoteproc.c
index 99c1e94c9366..c8822afab394 100644
--- a/drivers/remoteproc/andes_remoteproc.c
+++ b/drivers/remoteproc/andes_remoteproc.c
@@ -39,6 +39,7 @@ struct andes_rproc_pdata {
 	int num_local_mem_access_ports;
 	u64 mp_local_mem_access_port[2];
 	u64 sp_local_mem[2];
+	u64 local_mem_size[2];
 	int irq;
 	void __iomem *mbox_msg;
 	void __iomem *atcsmu_base;
@@ -242,6 +243,7 @@ static int andes_parse_local_mem_access_ports(struct rproc *rproc)
 
 			local->mp_local_mem_access_port[i] = rsc.start;
 			local->sp_local_mem[i] = size & rsc.start;
+			local->local_mem_size[i] = size;
 			rproc_add_carveout(rproc, mem);
 		}
 	}
@@ -325,8 +327,9 @@ int andes_rproc_elf_load_segments(struct rproc *rproc, const struct firmware *fw
 		}
 
 		for (j = 0; j < local->num_local_mem_access_ports; j++) {
-			if (da == local->sp_local_mem[j])
-				da = local->mp_local_mem_access_port[j];
+			if (da >= local->sp_local_mem[j] &&
+			    da < (local->sp_local_mem[j] + local->local_mem_size[j]))
+				da |= local->mp_local_mem_access_port[j];
 		}
 
 		/* grab the kernel address for this device address */
-- 
2.34.1

